Fixing BinarySearch section (#315)
NuevoFoundation/workshops: content/english/debugging/04-identify-the-problem3.md

## Debuggers

> Debugging code is such a widespread, common problem that people have built entire programs designed to help other programmers debug more efficiently. These are suitably named **debuggers**, and there are plenty of debuggers that work for the C programming language. Let's take a look at `gdb`, a common debugger used on the command line.

Reviewer: ```suggestion
Debugging code is such a widespread, common problem that people have built entire programs designed to help other programmers debug more efficiently. These are suitably named **debuggers**, and there are plenty of debuggers that work for the C programming language. Let's take a look at `gdb`, a common debugger used with the command line.
```

> [GDB (GNU Project Debugger)](https://sourceware.org/gdb/) is a powerful debugger that lets you debug programs from the command line, which is useful in cases where you don’t have access to a GUI. Open the **Shell** tab in the Repl.it window below. 

Reviewer: ```suggestion
[GDB (GNU Project Debugger)](https://sourceware.org/gdb/) is a powerful debugger that lets you debug programs from the command line, which is useful in cases where you don’t have access to a GUI. Open the **Shell** tab in the Replit window below. 
```

```

{{% notice note %}}
When the program is compiled, we use two flags: `-g` and `-Og`. The first tells the compiler to add **debugging information**, meaning that without that flag, gdb would not be able to debug the program. The second tells the compiler to optimize the program in a way that wouldn't affect the program's run structure. Without `-Og`, the compiler could potentially optimize away some of your code, so a debugger would not be nearly as effective.

> It is important to keep note of the second flag. For debugging, you should ALWAYS make sure the compiler makes minimal optimizations to your code because optimizations could drastically change how the code is run!

Reviewer: ```suggestion
It is important to keep note of the second flag. For debugging, you should **ALWAYS** make sure the compiler makes minimal optimizations to your code because optimizations could drastically change how the code is run!

```
